### PR TITLE
thread leakage checker and memory usage reporter #1226

### LINKED ISSUE #1226: Checking Resource Leak at end of a test class
Recently unit test can not complete due to threads leaking and ZK session leaking. We identified the leaks and fixed them. In order for future test not to leak resources, we need a way to enforce test class would release/clean up resource created during test.

A check can/should be implemented in @afterclass in ZkTestBase.java.

### PR DESCRIPTION
**Pull request by kaisun2000**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

fix #1226

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

Add thread leakage checker and memory usage reporter. The two
utility would be invoke before and after test classes. The would
help to detect/monitor resource/memory usage of the unit test.
### Tests

- [ ] The following tests are written for this issue:

(List the names of added unit/integration tests)

- [x] The following is the result of the "mvn test" command on the appropriate module:

2020-10-08T07:55:24.0931119Z 
2020-10-08T07:55:24.4922133Z [ERROR] Failures: 
2020-10-08T07:55:24.4923510Z [ERROR]   TestDisableCustomCodeRunner.test:236 expected:<false> but was:<true>
2020-10-08T07:55:24.4925729Z [ERROR]   TestClusterStatusMonitorLifecycle.testClusterStatusMonitorLifecycle:290 expected:<true> but was:<false>
2020-10-08T07:55:24.4928050Z [ERROR] Tests run: 1213, Failures: 2, Errors: 0, Skipped: 0
re-run would work.

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java [hunk 1]
```diff
@@ -127,8 +129,22 @@
   protected Map<String, ClusterSetup> _clusterSetupMap = new HashMap<>();
   protected Map<String, BaseDataAccessor> _baseDataAccessorMap = new HashMap<>();
 
+  static public void reportPhysicalMemory() {
+    com.sun.management.OperatingSystemMXBean os = (com.sun.management.OperatingSystemMXBean)
+        java.lang.management.ManagementFactory.getOperatingSystemMXBean();
+    long physicalMemorySize = os.getTotalPhysicalMemorySize();
+    System.out.println("************ SYSTEM Physical Memory:"  + physicalMemorySize);
+
+    int MB = 1024 * 1024;
```
Comment (lei-xia) [inline helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java:138]: minor: int -> long?

File: helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java [hunk 2]
```diff
@@ -710,7 +726,9 @@ protected Message createMessage(Message.MessageType type, String msgId, String f
   }
 
   @AfterClass
-  public void cleanupLiveInstanceOwners() {
+  public void cleanupLiveInstanceOwners() throws InterruptedException {
```
Comment (lei-xia) [inline helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java:729]: Will this method being called by all sub test classes?

**Comment by lei-xia** [comment]
Are we targeting to output the memory usage after each test class? Will this PR make it happen?

File: helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    ThreadGroup candidate = Thread.currentThread().getThreadGroup();
+    while (candidate.getParent() != null) {
+      candidate = candidate.getParent();
+    }
+    return candidate;
+  }
+
+  private static List<Thread> getAllThreads() {
+    ThreadGroup rootThreadGroup = getRootThreadGroup();
+    Thread[] threads = new Thread[32];
```
Comment (zhangmeng916) [inline helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java:27]: Isn't the 32 too small as initial value? We have a lot more than that right?

File: helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+
+  private static final String[] ZkServerThrdPattern =
+      {"SessionTracker", "NIOServerCxn", "SyncThread:", "ProcessThread"};
+  private static final String[] ZkSessionThrdPattern =
+      new String[]{"ZkClient-EventThread", "ZkClient-AsyncCallback", "-EventThread", "-SendThread"};
+  private static final String[] ForkJoinThrdPattern = new String[]{"ForkJoinPool"};
+  private static final String[] TimerThrdPattern = new String[]{"time"};
+  private static final String[] TaskStateModelThrdPattern = new String[]{"TaskStateModel"};
+
+  private static enum ThreadCategory {
```
Comment (zhangmeng916) [inline helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java:44]: Can you give some explanation of the default value? Otherwise, it's a bit hard to follow. E.g., whey ForkJoin is default to 2?

File: helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+      });
+    });
+
+    // step 3: enforce checking policy
+    boolean checkStatus = true;
+    for (ThreadCategory threadCategory : ThreadCategory.values()) {
+      int limit = threadCategory.getLimit();
+      int warningLimit = threadCategory.getWarningLimit();
+
+      Integer catThreadCnt = threadByCnt.get(threadCategory);
```
Comment (zhangmeng916) [inline helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java:148]: nit: can you use more character for "category", catThreadCnt is not quit self explanatory.

File: helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+        }
+        if (!dumpThread) {
+          continue;
+        }
+        // print first 10 thread names
+        int i = 0;
+        for (Thread t : threadByCat.get(threadCategory)) {
+          System.out.println(i + " thread:" + t.getName());
+          i++;
+          if (i == 100) {
```
Comment (zhangmeng916) [inline helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java:171]: Comments said "10", but I think "100" makes more sense.

File: helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java (hunk 2 above)
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java:729]: All the ones using ZkTestBase

File: helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java [hunk 7]
```diff
@@ -719,6 +737,17 @@ public void cleanupLiveInstanceOwners() {
       clientMap.clear();
     }
     _liveInstanceOwners.clear();
+
+    boolean status = false;
+    try {
+      status = ThreadLeakageChecker.afterClassCheck(testClassName);
```
Comment (zhangmeng916) [inline helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java:743]: Just to make sure, having this check here should cover all tests?

**Comment by kaisun2000** [comment]
> Are we targeting to output the memory usage after each test class? Will this PR make it happen?

Currently, all the ones using ZkTestBase. Since the majority use  ZkTestBase, we have a good idea of the memory usage over time.

File: helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java (hunk 4 above)
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java:44]: Good question. The two threshold are mostly empirical. 

ZkServer, our version has only 4 threads now. In case later version use more. I set the limit to 100. The reasoning is that these ZkServer threads are not deemed as leaking no matter how much they have.

ZkSession is the ZkClient and native Zookeeper client we have. ZkTestBase has 12 at starting up time. Thus, if there is more than that, it is the test code leaking ZkClient.

ForkJoin is created by using parallel stream or whatever Java features. This is out of our control, similar to ZkServer. Let me change the _limit to 100 while keep a small _warningLimit. (Will make change for this one).

Timer  should not happen. Setting limit to 2 not 0 mostly because even when you cancel the timer thread, it may take some not deterministic time for it to go away. So give it some slack here.

File: helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java (hunk 7 above)
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java:743]: Only those test derived from ZkTestBase. That is majority for now. We can open another ticket to track using then for all.

File: helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java (hunk 3 above)
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java:27]: It does not matter. Note in line 30 ` threads = new Thread[threads.length * 2];`, we double the threads array if they are not large enough, exponentially.

File: helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java (hunk 6 above)
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java:171]: changed the comments

File: helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java (hunk 5 above)
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java:148]: changed

File: helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java (hunk 1 above)
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java:138]: changed.

**Review by zhangmeng916 — APPROVED** [review]

**Comment by kaisun2000** [comment]
The PR is approved, please help to merge in.

>Add thread leakage checker and memory usage reporter. The two
utility would be invoke before and after test classes. The would
help to detect/monitor resource/memory usage of the unit test.

File: helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java [hunk 8]
```diff
@@ -0,0 +1,181 @@
+package org.apache.helix;
+
```
Comment (pkuwm) [inline helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java:2]: @kaisun2000 Can you add Apache License ?

File: helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java [hunk 9]
```diff
@@ -33,6 +33,7 @@
 import javax.management.MBeanServerConnection;
 import javax.management.ObjectName;
 
+import bsh.This;
```
Comment (pkuwm) [inline helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java:36]: Is this needed?

File: helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java (hunk 4 above)
Comment (pkuwm) [inline helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java:44]: I didn't mean to be picky, but I suggest we still follow the constant style: `static final String[] THIS_IS_CONSTANT`

File: helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java (hunk 8 above)
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java:2]: added.

File: helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java (hunk 9 above)
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java:36]: removed,

File: helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java (hunk 4 above)
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java]: changed.

File: helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import org.apache.helix.common.ZkTestBase;
+
+
+public class ThreadLeakageChecker {
```
Comment (jiajunwang) [inline helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java:35]: Please comment on what is this checker doing.

File: helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java (hunk 4 above)
Comment (jiajunwang) [reply helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java:63]: Please put the comment into the class. Code reviewers in the future shall not depend on this Github PR to understand the logic.

File: helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+    return Arrays.asList(Arrays.copyOf(threads, count));
+  }
+
+  private static final String[] ZKSERVER_THRD_PATTERN =
+      {"SessionTracker", "NIOServerCxn", "SyncThread:", "ProcessThread"};
+  private static final String[] ZKSESSION_THRD_PATTERN =
+      new String[]{"ZkClient-EventThread", "ZkClient-AsyncCallback", "-EventThread", "-SendThread"};
+  private static final String[] FORKJOIN_THRD_PATTERN = new String[]{"ForkJoinPool"};
+  private static final String[] TIMER_THRD_PATTERN = new String[]{"time"};
+  private static final String[] TASKSTATEMODEL_THRD_PATTERN = new String[]{"TaskStateModel"};
```
Comment (jiajunwang) [inline helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java:61]: Any thoughts on how we are going to maintain these hardcoded patterns?

File: helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java (hunk 10 above)
Comment (jiajunwang) [reply helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java:35]: Another question, does it only work for sequentially run tests?
If so, please note in the comment.

File: helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java [hunk 12]
```diff
@@ -710,7 +725,9 @@ protected Message createMessage(Message.MessageType type, String msgId, String f
   }
 
   @AfterClass
-  public void cleanupLiveInstanceOwners() {
+  public void cleanupLiveInstanceOwners() throws InterruptedException {
+    String testClassName = this.getShortClassName();
+    System.out.println("AfterClass:" + testClassName + " afterclass of ZkTestBase called!");
```
Comment (jiajunwang) [inline helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java:730]: Please remove this output. I think with no leakage found, we don't need this information.

File: helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java [hunk 13]
```diff
@@ -719,6 +736,17 @@ public void cleanupLiveInstanceOwners() {
       clientMap.clear();
     }
     _liveInstanceOwners.clear();
+
+    boolean status = false;
+    try {
+      status = ThreadLeakageChecker.afterClassCheck(testClassName);
+    } catch (Exception e) {
+      System.out.println("ThreadLeakageChecker exception:" + e.getStackTrace());
```
Comment (jiajunwang) [inline helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java:744]: Why system out print instead of log? If I redirect the test log into a file for debugging, I won't see this information.

File: helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiZkHelixJavaApis.java [hunk 14]
```diff
@@ -170,6 +171,9 @@ public void beforeClass() throws Exception {
 
   @AfterClass
   public void afterClass() throws Exception {
+    String testClassName = getClass().getSimpleName();
+    System.out.println("AfterClass: " + testClassName + " of TestMultiZkHelixJavaApis called.");
```
Comment (jiajunwang) [inline helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiZkHelixJavaApis.java:175]: Same comments as the ZkTestBase class.

File: helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java (hunk 11 above)
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java:61]: As long as developers start to use them, they can easier break out more categories and make the changes accordingly. Eventually the baton is passed code contributors.

File: helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java (hunk 10 above)
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java:35]: I think the checker by itself is thread safe. But currently if multiple test is running, checker won't be of any use for obvious reasons. Let me add an comment its usage depends on test running sequentially.

File: helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java (hunk 4 above)
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java:63]: added the comment

File: helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java (hunk 10 above)
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java:35]: added to the comment that we need to run tests sequentially to make this leakage check to be meaningful.

File: helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java (hunk 12 above)
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java:730]: removed

File: helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java (hunk 13 above)
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java:744]: Good point. changed to log.error

File: helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java (hunk 11 above)
Comment (jiajunwang) [reply helix-core/src/test/java/org/apache/helix/ThreadLeakageChecker.java:61]: How about new thread names or new threads that are not tracked by these patterns?
As a minimum requirement, I suggest tracking unknown threads as well. And a limitation should be added onto those threads as well. So if anyone changes the name, or introducing new threads that are leaking, the test will break. And they will be forced to modify this section.

**Review by jiajunwang — COMMENTED** [review]
Still, 2 problems that I also mentioned in the other test improvement PRs.
1. What is the guideline for using a system print vs. log? Without confirming this point, I cannot review it effectively. I prefer log, BTW.
2. Why not throw an exception on leakage? If just print error, then we can just let the tool runs every night. There is no need to do it with every "mvn test".
For example, if people ignore the error, then we might figure it out weeks later. Then this check will be very less helpful.

**Comment by kaisun2000** [comment]
> Still, 2 problems that I also mentioned in the other test improvement PRs.
> 
> 1. What is the guideline for using a system print vs. log? Without confirming this point, I cannot review it effectively. I prefer log, BTW.
> 2. Why not throw an exception on leakage? If just print error, then we can just let the tool runs every night. There is no need to do it with every "mvn test".
>    For example, if people ignore the error, then we might figure it out weeks later. Then this check will be very less helpful.

If we throw exception, current test won't work. I reduced the threads leakage from around 3000 to several hundreds at this point. Getting it to 0 is still some way to go.

**Comment by kaisun2000** [comment]
Sync-ed with JJ offline. Once the thread leakage reduced to 0. We will enable the throw exception (failing test). Before that, we use system.out to log it out for people to examine without enable info level logging

**Review by jiajunwang — APPROVED** [review]

**Comment by kaisun2000** [comment]
The PR is approved. Please help to merge to trunk

>Add thread leakage checker and memory usage reporter. The two
utility would be invoke before and after test classes. The would
help to detect/monitor resource/memory usage of the unit test.
